### PR TITLE
Release refinements

### PR DESCRIPTION
**Pull request by andschwa**
A number of release improvements, also fixes https://github.com/PowerShell/vscode-powershell/issues/2536.

### CODE REVIEW

File: tools/ReleaseTools.psm1 [hunk 1]
```diff
@@ -354,6 +372,8 @@ function New-DraftRelease {
         Pop-Location
     }
 
-    Get-GitHubRepository -OwnerName PowerShell -RepositoryName $RepositoryName |
-        New-GitHubRelease @ReleaseParams
+    $Repo = Get-GitHubRepository -OwnerName PowerShell -RepositoryName $RepositoryName
+    $Release = $Repo | New-GitHubRelease @ReleaseParams
+
+    $Assets | ForEach-Object { $Release | New-GitHubReleaseAsset -Path $_ }
```
Comment (andschwa) [inline tools/ReleaseTools.psm1:378]: @rjmholt Is this how you would do it all pipeline-y? Or do I need to set this up with begin/process/end?

File: docs/development.md [hunk 2]
```diff
@@ -80,17 +80,65 @@ New-DraftRelease -RepositoryName vscode-powershell
...
+`vX.Y.Z`. We do not release previews frequently because this dependency is not
+generally used directly: it's a library consumed by other projects which
+themselves use preview releases for beta testing.
+
+For the VS Code PowerShell Extension, our version follows `vYYYY.MM.X`, that is:
+current year, current month, and patch version (not day). This is not semantic
+versioning because of issues with how the VS Code marketplace and extension
+hosting API itself uses our version number. This scheme _does not_ mean we
+release on a chronological schedule: we release based on completed work. If the
+month has changed over since the last release, the patch version resets to 1.
```
Comment (rjmholt) [inline docs/development.md]: I thought we reset the patch version to 0

File: tools/ReleaseTools.psm1 [hunk 3]
```diff
@@ -332,28 +345,25 @@ function New-DraftRelease {
...
+        # NOTE: We rely on GitHub to create the tag.
         Tag        = "v$Version"
+        Committish = "release/v$Version"
         Name       = "v$Version"
         Body       = $ChangeLog
         PreRelease = [bool]$Version.PreReleaseLabel
         # TODO: Pass -WhatIf and -Confirm parameters correctly.
     }
 
-    if ($PSCmdlet.ShouldProcess("$RepositoryName/v$Version", "git tag")) {
```
Comment (rjmholt) [inline tools/ReleaseTools.psm1:346]: Looks like this command will no longer support shouldprocess, so need to take it out of the cmdletbinding

File: tools/ReleaseTools.psm1 [hunk 4]
```diff
@@ -332,28 +345,25 @@ function New-DraftRelease {
...
-        Push-Location -Path "$PSScriptRoot/../../$RepositoryName"
-        if (-not (git show-ref --tags "v$Version")) {
-            git tag "v$Version"
-        } else {
-            Write-Warning "git tag $RepositoryName/v$Version already exists!"
-        }
-        Pop-Location
-    }
+    $Repo = Get-GitHubRepository -OwnerName PowerShell -RepositoryName $RepositoryName
+    $Release = $Repo | New-GitHubRelease @ReleaseParams
```
Comment (rjmholt) [inline tools/ReleaseTools.psm1:366]: Piping to `New-GitHubRelease` will only take the URI from the input I think ([source](https://github.com/microsoft/PowerShellForGitHub/blob/0ebf9e8b8d0f4fa361996674dbca804da36bace7/GitHubReleases.ps1#L218-L378)).

It's easier to just to add OwnerName and RepositoryName to `$ReleaseParams` than to invoke `Get-GitHubRepository` I'd say.

File: tools/ReleaseTools.psm1 (hunk 1 above)
Comment (rjmholt) [reply tools/ReleaseTools.psm1:368]: It looks like [`New-GitHubReleaseAsset` takes the value of `-Path` from the pipeline](https://github.com/microsoft/PowerShellForGitHub/blob/0ebf9e8b8d0f4fa361996674dbca804da36bace7/GitHubReleases.ps1#L1024-L1030), so I'd write this like:

```powershell
$Assets | New-GitHubReleaseAsset -OwnerName PowerShell -RepositoryName $RepositoryName -Release $Release.id
```

File: docs/development.md [hunk 5]
```diff
@@ -80,17 +80,65 @@ New-DraftRelease -RepositoryName vscode-powershell
 # Upload PowerShellEditorServices.zip (for other extensions)
 # Upload VSIX and Install-VSCode.ps1
 # Publish draft releases and merge (don't squash!) branches
+# Check telemetry for stability before releasing
 vsce publish --packagePath ./PowerShell-<version>.vsix
 # Update Install-VSCode.ps1 on gallery
 Publish-Script -Path ./Install-VSCode.ps1 -NuGetApiKey (Get-Secret "PowerShell Gallery API Key" -AsPlainText)
 ```
 
+### Versioning
```
Comment (rjmholt) [inline docs/development.md:89]: This section is so great!

File: tools/ReleaseTools.psm1 [hunk 6]
```diff
@@ -332,28 +345,25 @@ function New-DraftRelease {
         [ValidateSet([RepoNames])]
         [string]$RepositoryName
     )
+    param(
+        [Parameter(ValueFromPipeline)]
+        [string[]]$Assets
+    )
     $Version = Get-Version -RepositoryName $RepositoryName
     $Changelog = (Get-FirstChangelog -RepositoryName $RepositoryName) -join "`n"
     $ReleaseParams = @{
```
Comment (rjmholt) [inline tools/ReleaseTools.psm1:354]: I notice all the local variable names here are in PascalCase, but usually for local variables I'd use camelCase

File: tools/ReleaseTools.psm1 (hunk 1 above)
Comment (andschwa) [reply tools/ReleaseTools.psm1:368]: Can we send `$Repo` and `$Assets` at the same time?

File: tools/ReleaseTools.psm1 (hunk 4 above)
Comment (andschwa) [reply tools/ReleaseTools.psm1:366]: Fair.

File: docs/development.md (hunk 5 above)
Comment (andschwa) [reply docs/development.md:89]: Thanks!

File: tools/ReleaseTools.psm1 (hunk 6 above)
Comment (andschwa) [reply tools/ReleaseTools.psm1:352]: 🤷

**Comment by andschwa** [comment]
@rjmholt After this is merged in the morning, I'll release a hot-fix for the stable channel that has all the files correctly signed.

**Review by rjmholt — APPROVED** [review]

**Pull request merged**